sam.c

```diff
@@ -759,6 +774,7 @@ static int fixup_missing_qname_nul(bam1_t *b) {
  * Note a second interface that returns a bam pointer instead would avoid bam_copy1
  * in multi-threaded handling.  This may be worth considering for htslib2.
  */
+#define bgzf_read bgzf_read_small
```

Comment:
I'd previously tested merging the 4 byte and 32 byte read together and concluded it made no difference in speed (once I'd put the inline version in - it did do before that), and it meant more changes which would have made reviewing more work.

I can try it again if you want, but it's more changes to evaluate.

Reply:
Also, I don't get the comment about `bgzf_read(fp, x, 32)` remaining as-is.  Initially it wasn't because it's small enough to be inlined and a fixed 32-byte read may still cause the memcpy to be optimised away (I don't know), and because we weren't subverting things and copying out the struct directly initially.  However since then, it may not matter much either way as that code path will be very rare.  I think it's probably an irrelevance at best.

Reply:
You also have to think about what's actually happening here.  Why is it that `bgzf_read(fp, x, 4)` for example is faster with the `bgzf_read_fast` inline?  Some of it is the removal of a function, but I think a lot of it is because when we inline a memcpy of 4 bytes it becomes a straight 32-bit load to register.  When we call the function, the length is a variable so it has to call memcpy direct and that in turn has a lot of code in it for dealing with arbitrary length and is probably optimised for copying long things more than short things.

So anywhere we have a `bgzf_read` with a small-ish constant size, it's generally good practice to call the inline variant so the compiler can optimise away memcpy.

Reply:
The optimisation can never work in the specific case of `bgzf_read_small(fp, x, 32)` due to the earlier line:
```c
if (fp->block_length - fp->block_offset > 32) {
```
i.e. the same optimisation (plus a zero-copy one) happens elsewhere.

Reply:
Ah yes I see what you mean.  It makes no difference to CPU from what I can see (maybe the compiler detects it as an impossibility and removes the code?), but yes it's pointless to go via the static inline.

Now we have the same code for both paths in parsing the `x[]` array/pointer there's no as much gain to be had on the `fp->block_length - fp->block_offset > 32` check anyway (if we kept the "small" variant), however it does still seem to be about 5% faster on small small-read datasets so we may as well keep it.